File: golang/graph_dfs/number_of_provinces.go
package main

func findCircleNum(isConnected [][]int) int {
	n := len(isConnected)
	visited := make([]bool, n)
	provinces := 0

	var dfs func(int)

	dfs = func(city int) {
		visited[city] = true

		for neighbor := 0; neighbor < n; neighbor++ {
			if isConnected[city][neighbor] == 1 && !visited[neighbor] {
				dfs(neighbor)
			}
		}
	}

	for i := 0; i < n; i++ {
		if !visited[i] {
			provinces++
			dfs(i)
		}
	}

	return provinces
}

File: java/src/main/java/com/sanedge/interview/graph_dfs/NumberOfProvinces.java
package com.sanedge.interview.graph_dfs;

public class NumberOfProvinces {

  public int findCircleNum(int[][] isConnected) {
    int n = isConnected.length;

    boolean[] visited = new boolean[n];

    int provinces = 0;

    for (int i = 0; i < n; i++) {
      if (!visited[i]) {
        provinces++;
        dfs(i, isConnected, visited);
      }
    }

    return provinces;
  }

  private void dfs(int city, int[][] isConnected, boolean[] visited) {
    visited[city] = true;

    for (int neighbor = 0; neighbor < isConnected.length; neighbor++) {
      if (isConnected[city][neighbor] == 1 && !visited[neighbor]) {
        dfs(neighbor, isConnected, visited);
      }
    }

  }
}
